exquisite_corpus/spm_naive_language_model.py

```diff
@@ -0,0 +1,428 @@
+import click
+import json
+import numpy as np
+import sys
+import time
+import torch
+import torch.nn as nn
+
+from exquisite_corpus.spm_dataset import SpmIdsBatchSampler, SpmIdsDataset
+from pathlib import Path
+from torch.utils.data import DataLoader
+
+
+class LanguageModel(nn.Module):
+    """
+    Since this is just to illustrate how one might set up a language model
+    using SentencePiece ids, we use a very simple model rather than a good one.
+    """
+
+    def __init__(
+        self,
+        n_tokens,
+        n_dims=300,
+        hidden_size=500,
+        num_layers=1,
+        bias=True,
+        dropout=0.0,
+        bidirectional=False,
+        sparse=False,
+    ):
+        super().__init__()
+        self.encoder = nn.Embedding(n_tokens, n_dims, sparse=bool(sparse))
+        self.rnn = nn.GRU(
+            input_size=n_dims,
+            hidden_size=hidden_size,
+            num_layers=num_layers,
+            bias=bool(bias),
+            batch_first=True,
+            dropout=dropout,
+            bidirectional=bool(bidirectional),
+        )
+        n_rnn_out_features = hidden_size * (2 if bidirectional else 1)
+        self.decoder = nn.Linear(in_features=n_rnn_out_features, out_features=n_tokens)
+        self.log_softmax = nn.LogSoftmax(dim=2)
+
+    def forward(self, input, hidden_states=None, return_hidden=False):
+        batch_size = input.size(0 if self.rnn.batch_first else 1)
+        n_directions = 2 if self.rnn.bidirectional else 1
+        full_num_layers = self.rnn.num_layers * n_directions
+        if hidden_states is None:
+            hidden_states = torch.zeros(
+                full_num_layers, batch_size, self.rnn.hidden_size
+            ).to(input.device, non_blocking=True)
+        output = input
+        output = self.encoder(output)
+        output, new_hidden_states = self.rnn(
+            output,
+            hidden_states.view(full_num_layers, batch_size, self.rnn.hidden_size),
+        )
+        output = self.decoder(output)
+        output = self.log_softmax(output)
+        if return_hidden:
+            new_hidden_states = new_hidden_states.view(hidden_states.size())
+            return output, new_hidden_states
+        else:
+            return output
+
+
+class LossFunction(nn.Module):
+    """
+    Module to compute the aggregate (over all elements of a batch, and all
+    sequence positions within each element) loss.
+    """
+
+    def __init__(self, reduction="mean"):
+        super().__init__()
+        self.loss_fn = nn.NLLLoss(reduction=reduction)
+
+    def forward(self, input, target):
+        return self.loss_fn(input.view(target.numel(), -1), target.view(-1))
```

Comment:
Add comments explaining what is happening

Reply:
Done.